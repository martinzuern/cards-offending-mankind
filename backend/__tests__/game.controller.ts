import request from 'supertest';
import { v4 as uuidv4 } from 'uuid';
import ServerInstance from '../server';
import type { Application } from 'express';

let App: Application;
let Teardown: () => Promise<void>;

beforeAll(async (done) => {
  const { app, closeServer } = await ServerInstance;
  App = app;
  Teardown = closeServer;
  done();
});

afterAll((done) => Teardown().then(() => done()), 90000);

describe('Game without password', () => {
  let createdGame;

  describe('create', () => {
    it('should return error if missing params', (done) =>
      request(App)
        .post('/api/v1/games')
        .send({})
        .expect('Content-Type', /json/)
        .expect(400)
        .expect((r) => {
          expect(r.body).toMatchSnapshot();
        })
        .end(done));

    it('should return new game', (done) =>
      request(App)
        .post('/api/v1/games')
        .send({ player: { nickname: 'foo' }, game: {} })
        .expect('Content-Type', /json/)
        .expect(201)
        .expect((r) => {
          expect(r.body).toMatchSnapshot({
            game: {
              id: expect.any(String),
            },
            player: {
              id: expect.any(String),
              token: expect.any(String),
            },
          });
          createdGame = r.body;
        })
        .end(done));
  });

  describe('get', () => {
    it('renders 400 if not valid uuidv4', (done) =>
      request(App)
        .get('/api/v1/games/foo')
        .expect('Content-Type', /json/)
        .expect(400)
        .expect((r) => {
          expect(r.body).toMatchSnapshot();
        })
        .end(done));

    it('renders 404 if not found', (done) =>
      request(App)
        .get(`/api/v1/games/${uuidv4()}`)
        .expect('Content-Type', /json/)
        .expect(404)
        .expect((r) => {
          expect(r.body).toMatchSnapshot();
        })
        .end(done));

    it('allows to get game', (done) =>
      request(App)
        .get(`/api/v1/games/${createdGame.game.id}`)
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((r) => {
          expect(r.body).toEqual({ game: createdGame.game });
        })
        .end(done));
  });

  describe('join', () => {
    it('barfs if no nickname given', (done) =>
      request(App)
        .post(`/api/v1/games/${createdGame.game.id}/join`)
        .send({})
        .expect('Content-Type', /json/)
        .expect(400)
        .expect((r) => {
          expect(r.body).toMatchSnapshot();
        })
        .end(done));

    it('barfs if nickname already in use', (done) =>
      request(App)
        .post(`/api/v1/games/${createdGame.game.id}/join`)
        .send({ nickname: 'foo' })
        .expect('Content-Type', /json/)
        .expect(400)
        .expect((r) => {
          expect(r.body).toMatchSnapshot();
        })
        .end(done));

    it('allows to join game', (done) =>
      request(App)
        .post(`/api/v1/games/${createdGame.game.id}/join`)
        .send({ nickname: 'foo2' })
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((r) => {
          expect(r.body).toMatchSnapshot({
            player: {
              id: expect.any(String),
              token: expect.any(String),
            },
          });
        })
        .end(done));
  });
});

describe('Game with password', () => {
  const password = 'password123';
  let createdGame;

  describe('create', () => {
    it('should return new game', (done) =>
      request(App)
        .post('/api/v1/games')
        .send({ player: { nickname: 'foo' }, game: { password } })
        .expect('Content-Type', /json/)
        .expect(201)
        .expect((r) => {
          expect(r.body).toMatchSnapshot({
            game: {
              id: expect.any(String),
            },
            player: {
              id: expect.any(String),
              token: expect.any(String),
            },
          });
          createdGame = r.body;
          done();
        })
        .end(done));
  });

  describe('get', () => {
    it('allows to get game', (done) =>
      request(App)
        .get(`/api/v1/games/${createdGame.game.id}`)
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((r) => {
          expect(r.body).toEqual({ game: createdGame.game });
        })
        .end(done));
  });

  describe('join', () => {
    it('barfs if no password given', (done) =>
      request(App)
        .post(`/api/v1/games/${createdGame.game.id}/join`)
        .send({ nickname: 'foo2' })
        .expect('Content-Type', /json/)
        .expect(403)
        .expect((r) => {
          expect(r.body).toMatchSnapshot();
        })
        .end(done));

    it('barfs if wrong password given', (done) =>
      request(App)
        .post(`/api/v1/games/${createdGame.game.id}/join`)
        .send({ nickname: 'foo2', password: 'wrong' })
        .expect('Content-Type', /json/)
        .expect(403)
        .expect((r) => {
          expect(r.body).toMatchSnapshot();
        })
        .end(done));

    it('allows to join game', (done) =>
      request(App)
        .post(`/api/v1/games/${createdGame.game.id}/join`)
        .send({ nickname: 'foo2', password })
        .expect('Content-Type', /json/)
        .expect(200)
        .expect((r) => {
          expect(r.body).toMatchSnapshot({
            player: {
              id: expect.any(String),
              token: expect.any(String),
            },
          });
        })
        .end(done));
  });
});
