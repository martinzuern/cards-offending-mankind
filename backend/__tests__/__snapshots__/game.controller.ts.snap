// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game with password create should return new game 1`] = `
Object {
  "game": Object {
    "handSize": 10,
    "hasPassword": true,
    "id": Any<String>,
    "packs": Array [],
    "specialRules": Array [],
    "status": "created",
    "timeouts": Object {
      "betweenRounds": 30,
      "judging": 120,
      "playing": 120,
      "revealing": 60,
    },
    "winnerPoints": 20,
  },
  "player": Object {
    "deck": Array [],
    "id": Any<String>,
    "isAI": false,
    "isActive": true,
    "isHost": true,
    "nickname": "foo",
    "points": 0,
    "token": Any<String>,
  },
}
`;

exports[`Game with password join allows to join game 1`] = `
Object {
  "player": Object {
    "deck": Array [],
    "id": Any<String>,
    "isAI": false,
    "isActive": true,
    "isHost": false,
    "nickname": "foo2",
    "points": 0,
    "token": Any<String>,
  },
}
`;

exports[`Game with password join barfs if no password given 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Error: Game password not provided.",
    },
  ],
}
`;

exports[`Game with password join barfs if wrong password given 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Error: Game password incorrect",
    },
  ],
}
`;

exports[`Game without password create should return error if missing params 1`] = `
Object {
  "errors": Array [
    Object {
      "errorCode": "required.openapi.validation",
      "message": "should have required property 'game'",
      "path": ".body.game",
    },
    Object {
      "errorCode": "required.openapi.validation",
      "message": "should have required property 'player'",
      "path": ".body.player",
    },
  ],
}
`;

exports[`Game without password create should return new game 1`] = `
Object {
  "game": Object {
    "handSize": 10,
    "hasPassword": false,
    "id": Any<String>,
    "packs": Array [],
    "specialRules": Array [],
    "status": "created",
    "timeouts": Object {
      "betweenRounds": 30,
      "judging": 120,
      "playing": 120,
      "revealing": 60,
    },
    "winnerPoints": 20,
  },
  "player": Object {
    "deck": Array [],
    "id": Any<String>,
    "isAI": false,
    "isActive": true,
    "isHost": true,
    "nickname": "foo",
    "points": 0,
    "token": Any<String>,
  },
}
`;

exports[`Game without password get renders 400 if not valid uuidv4 1`] = `
Object {
  "errors": Array [
    Object {
      "errorCode": "format.openapi.validation",
      "message": "should match format \\"uuid\\"",
      "path": ".params.GameID",
    },
  ],
}
`;

exports[`Game without password get renders 404 if not found 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Resource not found.",
    },
  ],
}
`;

exports[`Game without password join allows to join game 1`] = `
Object {
  "player": Object {
    "deck": Array [],
    "id": Any<String>,
    "isAI": false,
    "isActive": true,
    "isHost": false,
    "nickname": "foo2",
    "points": 0,
    "token": Any<String>,
  },
}
`;

exports[`Game without password join barfs if nickname already in use 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Error: Nickname already taken",
    },
  ],
}
`;

exports[`Game without password join barfs if no nickname given 1`] = `
Object {
  "errors": Array [
    Object {
      "errorCode": "required.openapi.validation",
      "message": "should have required property 'nickname'",
      "path": ".body.nickname",
    },
  ],
}
`;
